Add in support for hilbert index as a part of zorder work

revans2 opened the pull request:
This is related to the interleave bits work done for zorder before. This is work to support the databricks internal implementation for zorder.

revans2 commented:
build

hyperbolic2346 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+  long_backed_array(long_backed_array const&) = default;  ///< Copy constructor
+  long_backed_array(long_backed_array&&) = default;  ///< Move constructor
+  inline __device__ explicit long_backed_array(int32_t num_bits): data(0), 
+    num_bits(num_bits),  mask(static_cast<uint64_t>((1L << num_bits) - 1)) {}
```
Should `num_bits` be named something more like `num_bits_per_entry` or some such? It took me a while to make the connection that it wasn't the total number of bits in the array. I also don't see any verification that the number of bits desired per element * number of elements is < sizeof(long).

revans2 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+// With thanks also to Paul Chernoch who published a C# algorithm for Skilling's
+// work on StackOverflow and
+// <a href="https://github.com/paulchernoch/HilbertTransformation">GitHub</a>.
+__device__ uint64_t to_hilbert_index(const long_backed_array & transposed_index, const int num_bits, const int num_dimensions) {
```
It is the same `num_bits`. I called it this to match closely with the original code.

https://github.com/davidmoten/hilbert-curve/blob/758058c0b86ff01edffb46317f04a619d861ba14/src/main/java/org/davidmoten/hilbert/HilbertCurve.java#L26

but I added `num_` to it because I thought `bits` was not descriptive enough. Happy to make it clearer.

revans2 commented:
build

ttnghia commented on src/main/cpp/src/ZOrderJni.cpp:
```diff
@@ -35,4 +35,20 @@ Java_com_nvidia_spark_rapids_jni_ZOrder_interleaveBits(JNIEnv *env, jclass, jlon
 }
 
+JNIEXPORT jlong JNICALL
+Java_com_nvidia_spark_rapids_jni_ZOrder_hilbertIndex(JNIEnv *env, jclass, jint num_bits, jlongArray input_columns) {
```
Why not pass in a table from Java instead?

revans2 replied:
It is minor but this reduces the number of JNI calls and in my opinion fits better with how Spark uses it. If we pass a table down from java then I have to create a table in java, which will require a JNI call.  I use the table internally to verify that the length of the columns are all the same, and so I can reuse the existing device_table_view.  Why does java need to know that is how I am doing it.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+// the data in a long with a set number of bits allocated for
+// each item (num_bits_per_entry)
+struct long_backed_array {
+  long_backed_array() = delete;
```
If we declare `data_type` inside this struct, we should call `long_backed_array::data_type` all the time. Otherwise, we should put it somewhere else (probably above this struct).

hyperbolic2346 replied:
That is the gist of the change. I was so tempted to create ```
template <typename T>
struct backed_array {
}```

I worry I'm too far in the weeds though and missing the point.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+// each item (num_bits_per_entry)
+struct long_backed_array {
+  long_backed_array() = delete;
+  ~long_backed_array() = default;
+  long_backed_array(long_backed_array const&) = default;  ///< Copy constructor
+  long_backed_array(long_backed_array&&) = default;  ///< Move constructor
```
I believe that you don't have to declare these default methods. Only declare what you need.

ttnghia replied:
They are just to dilute your code. Except for the overloaded constructor (you provide one then the compiler will not provide the default constructor), other methods are implicitly generated the same as you write `default`. So explicitly writing them `... = default` is redundant.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+  inline __device__ explicit long_backed_array(int32_t num_bits_per_entry): data(0),
+    num_bits_per_entry(num_bits_per_entry),  mask(static_cast<uint64_t>((1L << num_bits_per_entry) - 1)) {}
+
+  /**
+   * @brief Copy assignment operator
+   *
+   * @return Reference to this object
+   */
+  long_backed_array& operator=(long_backed_array const&) = default;
+  /**
+   * @brief Move assignment operator
+   *
+   * @return Reference to this object (after transferring ownership)
+   */
+  long_backed_array& operator=(long_backed_array&&) = default;
```
Similar to above.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+  ~long_backed_array() = default;
+  long_backed_array(long_backed_array const&) = default;  ///< Copy constructor
+  long_backed_array(long_backed_array&&) = default;  ///< Move constructor
+  inline __device__ explicit long_backed_array(int32_t num_bits_per_entry): data(0),
```
Also we can avoid redundant `inline` keyword. When a function only has definition (implemented without a separate declaration) then it is implicitly inline. So all the `inline` keyword in this file can be removed.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+   */
+  long_backed_array& operator=(long_backed_array&&) = default;
+
+  inline __device__ uint32_t operator[](int i) const {
+    int32_t offset = num_bits_per_entry * i;
```
Please try to enforce consistency: use `int32_t` all the time and avoid `int`.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -114,4 +226,53 @@ std::unique_ptr<cudf::column> interleave_bits(
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr) {
+ 
+  auto num_rows = tbl.num_rows();
+  auto num_columns = tbl.num_columns();
```
Use `const` as much as possible.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -114,4 +226,53 @@ std::unique_ptr<cudf::column> interleave_bits(
+                [](cudf::column_view const& col) { return col.type().id() == cudf::type_id::INT32; }),
+    "All columns of the input table must be INT32.");
+
+  auto input_dv = cudf::table_device_view::create(tbl, stream);
```
Also `auto const`.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -114,4 +226,53 @@ std::unique_ptr<cudf::column> interleave_bits(
+         row.set(column_index, data);
+       }
+
+       auto transposed_index = hilbert_transposed_index(row, num_bits_per_entry, num_columns);
```
`auto const`.

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -114,4 +226,53 @@ std::unique_ptr<cudf::column> interleave_bits(
+     num_bits_per_entry,
+     num_columns,
+     input = *input_dv] __device__ (cudf::size_type row_index) {
+       long_backed_array row(num_bits_per_entry);
```
`const`.

revans2 replied:
We modify the row. Why would I mark it as `const` to say that we cannot modify it?

ttnghia commented on src/main/cpp/src/zorder.cu:
```diff
@@ -114,4 +226,53 @@ std::unique_ptr<cudf::column> interleave_bits(
+
+  auto output_dv_ptr = cudf::mutable_column_device_view::create(*output_data_col, stream);
+
+  thrust::for_each_n(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<cudf::size_type>(0),
+    num_rows,
```
Nit: This can be rewritten using `thrust::transform`.

ttnghia commented on src/main/cpp/src/zorder.hpp:
```diff
@@ -29,4 +29,10 @@ std::unique_ptr<cudf::column> interleave_bits(
+std::unique_ptr<cudf::column> hilbert_index(
+  int32_t const num_bits,
+  cudf::table_view const& tbl,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
```
I think all the default stream should be accessed through `cudf::default_stream_value`.

ttnghia commented on src/main/java/com/nvidia/spark/rapids/jni/ZOrder.java:
```diff
@@ -54,5 +54,34 @@ public static ColumnVector interleaveBits(int numRows, ColumnVector ... inputCol
+  /**
+   * Yes a hilbertIndex is not technically used in zorder, but it is an alternative way to
+   * cluster the data that databricks uses, and that is why we have it here. Please note that
+   * this currently only supports indexes where numBits * inputColumns.length <= 64.
```
```suggestion
   * this currently only supports indexes where numBits * inputColumns.length <= 64.
   
```

ttnghia commented on src/main/java/com/nvidia/spark/rapids/jni/ZOrder.java:
```diff
@@ -54,5 +54,34 @@ public static ColumnVector interleaveBits(int numRows, ColumnVector ... inputCol
+   */
+  public static ColumnVector hilbertIndex(int numBits, int numRows, ColumnVector ... inputColumns) {
+    if (inputColumns.length == 0) {
+      try (Scalar zero = Scalar.fromLong(0)) {
+        return ColumnVector.fromScalar(zero, numRows);
```
Since the column returned from cuda code is `UINT64` type, this should also be `fromUnsignedLong`.

revans2 replied:
Like I said. Spark/Java do not use/support `uint64`. The value returned was computed using a uint64_t to avoid shift issues, but in order to match exactly what the java code does will be returned as an `int64`/`long`.

revans2 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+
+  inline __device__ void set(int i, uint32_t value) {
+    int32_t offset = i * num_bits_per_entry;
+    uint64_t masked_data = data & ~(static_cast<uint64_t>(mask) << offset);
```
I don't think we can drop the static_cast. I need to be able to shift the mask above what a 32-bit value can hold.

revans2 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,118 @@
+// With thanks also to Paul Chernoch who published a C# algorithm for Skilling's
+// work on StackOverflow and
+// <a href="https://github.com/paulchernoch/HilbertTransformation">GitHub</a>.
+__device__ uint64_t to_hilbert_index(const long_backed_array & transposed_index, const int num_bits_per_entry, const int num_dimensions) {
```
I would prefer to keep the uint64_t here separate from the data_type in the long_backed_array. Conceptually they are different types that are not related to each other.

revans2 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -114,4 +226,53 @@ std::unique_ptr<cudf::column> interleave_bits(
+  auto input_dv = cudf::table_device_view::create(tbl, stream);
+
+  auto output_data_col = cudf::make_numeric_column(
+      cudf::data_type{cudf::type_id::INT64}, num_rows, cudf::mask_state::UNALLOCATED, stream, mr);
```
I want an `INT64` so it matches what Spark expects.  Java does not have uint types so I want the output to actually be an int.

revans2 commented:
@hyperbolic2346 and @ttnghia I think I have addressed or replied to all of your comments. There were a few `const` that caused compile errors, but I think I fixed most of it. Please take another look.

revans2 commented:
build

revans2 commented:
build

hyperbolic2346 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,105 @@
+  }
+
+private:
+  uint64_t data;
```
```suggestion
  data_type data;
```

hyperbolic2346 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -27,6 +27,105 @@
+// <a href="https://github.com/paulchernoch/HilbertTransformation">GitHub</a>.
+__device__ uint64_t to_hilbert_index(long_backed_array const & transposed_index,
+        int32_t const num_bits_per_entry, int32_t const num_dimensions) {
+  uint64_t b = 0;
```
these are sized to match the `long_backed_array`, correct? I would expect these to be
`long_backed_array::data_type` instead of `uint64_t`. It is a bit pedantic, but if the backing store changes, it would be easy to miss changing these.

revans2 replied:
They are but also they are not. I picked the size of long_backed_array because of the size of the output needed in this case. So for me you are pushing for the dependency to go in the wrong direction. The size of long_backed_array is set by the desired output size, not the other way around.  Perhaps I should just change it to be templated. But I don't really want to because what we are trying to replace does not have more options than a long.

hyperbolic2346 reviewed (changes requested):
I'm ok with this change, but have a couple of questions.

revans2 commented:
build
